src/libAtomVM/nifs.c

```diff
@@ -3302,3 +3364,159 @@ static term nif_maps_next(Context *ctx, int argc, term argv[])
 
     return ret;
 }
+
+#ifndef AVM_NO_FP
+
+typedef avm_float_t (*unary_math_f)(avm_float_t x);
+typedef avm_float_t (*binary_math_f)(avm_float_t x, avm_float_t y);
+
+static void maybe_clear_exceptions()
+{
+#if defined(math_errhandling) && defined(MATH_ERREXCEPT)
+    if (math_errhandling & MATH_ERREXCEPT) {
+        feclearexcept(FE_ALL_EXCEPT);
+    }
+#endif
+}
+
+static term get_exception(Context *ctx)
+{
+    UNUSED(ctx);
```

Comment:
I guess at this point I can get rid of ctx as a parameter.  I think I may have thought I needed it in a previous revision.

Reply:
I don't know how this can work with SMP. This code may be executed on the other core.

Reply:
The C reference docs seem to suggest the exception handlers (data) are thread-local.

https://en.cppreference.com/w/c/numeric/fenv

Reply:
They are, I was confused by having a function with no parameter that obviously checks a global (thread local or not). With the SMP build, a process can run be scheduled on another thread when it is scheduled out, so we need to make sure this function is called before the process is scheduled out.

Reply:
Are you concerned that the process will be scheduled out by the (RT)OS, or by the AtomVM scheduler?

All of this code is being called in the context of a single Nif call, so I don’t think there is an opportunity for the AtomVM scheduler to pre-empt the call.  (No dirty schedulers in AtomVM — yet)

I am less certain of the OS context switch case, but I would think that the (RT)OS would preserve the state of any thread-local data if the thread/task were pre-empted.

Reply:
I realize my initial comment was both confusing and based on my own confusion, I probably shouldn't comment on code late in the night. Sorry for that.

I am concerned that the process may be scheduled out by the AtomVM scheduler, not in your current implementation because as you write, all this code is being called in the context of a single Nif call, but in a future usage of these functions. I was (and still am) reacting to your initial comment where you considered dropping the ctx argument.

Possible actions are:
- ignore my comments as your code is safe and proceed to remove the ctx argument
- find a way to signal somehow the usage (it should be called in the context of a single Nif call), for example replacing them with macros, static inline functions.

Reply:
Added a maintenance note to the code.